distributed/core.py

```diff
@@ -663,15 +663,25 @@ async def send_recv(comm, reply=True, serializers=None, deserializers=None, **kw
             response = await comm.read(deserializers=deserializers)
         else:
             response = None
-    except OSError:
+    # Note: this is asyncio.TimeoutError, not the builtin TimeoutError which is a
+    # subclass of OSError
+    except (TimeoutError, OSError):
         # On communication errors, we should simply close the communication
+        # This includes CommClosedError and socket timeouts
         force_close = True
         raise
+    except (CancelledError, asyncio.CancelledError):
+        # Do not reuse the comm to prevent the next call of send_recv from receiving
+        # data from this call and/or accidentally putting multiple waiters on read().
+        # Note that this relies on all Comm implementations to allow a write() in the
+        # middle of a read().
+        please_close = True
+        raise
     finally:
-        if please_close:
-            await comm.close()
-        elif force_close:
+        if force_close:
             comm.abort()
+        elif please_close:
+            await comm.close()
```

Comment:
In case of OSError, this should halve the latency of ``scheduler.broadcast`` (which is the only method that uses close=True).